lib/cylc/tests/__init__.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python2
+
+# THIS FILE IS PART OF THE CYLC SUITE ENGINE.
+# Copyright (C) 2008-2018 NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import doctest
+import os
+import pkgutil
+import sys
+import unittest
+
+
+def iter_package(name, path):
+    """Yield modules contained within the provided package."""
+    stack = [(name, path)]
+    while stack:
+        namespace, path = stack.pop()
+        for _, name, is_package in pkgutil.walk_packages([path]):
+            modname = '%s.%s' % (namespace, name)
+            modpath = os.path.join(path, name)
+            yield modname, modpath
+            if is_package:
+                stack.append((modname, modpath))
+
+
+def load_tests(loader, tests, _):
+    """Called by unittest to determine which tests to run.
+
+    See: https://docs.python.org/3/library/unittest.html#load-tests-protocol
+    """
+    import importlib
+
+    cylc_package_path = os.path.split(os.path.split(__file__)[0])[0]
+    for module_name, _ in iter_package('cylc', cylc_package_path):
+        # ensure the module is importable
+        try:
+            importlib.import_module(module_name)
+        except ImportError as exc:
+            print 'doctest: skipping "%s" - %s' % (module_name, exc)
```

Comment:
Is this just for user diagnostics?
* If so, should probably go to STDERR? And should we use the print function instead of the the print statement?
* Is there a better way to alert users that we are skipping tests? (Like we have in TAP format?)

Reply:
* This message is of the form "skipped python module not available" which is roughly comparable to the "skipped: command not available" messages we get from the TAP test harness. In TAP these messages are printed to stdout no stderr. Can convert to print functions if desired.
* In the unittest framework tests can be skipped but they must first exist in order to be skipped. The test suite is created here: https://github.com/cylc/cylc/pull/2766/files/6b3d4a82420a82e0424f8b79403708d9ce39374d#diff-d15de699c9b87f6274e3824b0b02abb2R56

Reply:
@oliver-sanders I think that code can now be deprecated, as the `doctest`s are now running with PyTest.

Checked running `py.test` with and without this code, and the result was the same.

Is it something that we need to keep for Python 2.6 compatibility? Just committed the change removing it to trigger Travis and check the coverage/test reports, but happy to revert if necessary 👍

Reply:
Always remove if not necessary!

> "I hate code, and I want as little of it as possible in our product." - Jack Diederich

So long as py.test picks up any and all doctests that's great!